src/misc/secret-doors.ts

```diff
@@ -0,0 +1,162 @@
+import { Bukkit, Material, Sound, SoundCategory } from 'org.bukkit';
+import { Block, BlockFace } from 'org.bukkit.block';
+import { Half } from 'org.bukkit.block.data.Bisected';
+import { Door } from 'org.bukkit.block.data.type';
+import { Hinge } from 'org.bukkit.block.data.type.Door';
+import { Player } from 'org.bukkit.entity';
+import { Action, BlockBreakEvent } from 'org.bukkit.event.block';
+import { PlayerInteractEvent } from 'org.bukkit.event.player';
+import * as yup from 'yup';
+import { CustomBlock } from '../common/blocks/CustomBlock';
+import { chanceOf } from '../common/helpers/math';
+
+const CobblestoneDoor = new CustomBlock({
+  type: Material.COBBLESTONE,
+  data: {
+    hasLeftHinge: yup.bool().required(),
+    facing: yup.string().required(),
+  },
+});
+const StoneBrickDoor = new CustomBlock({
+  type: Material.STONE_BRICKS,
+  data: {
+    hasLeftHinge: yup.bool().required(),
+    facing: yup.string().required(),
+  },
+});
+
+const DOORS = new Map([
+  [Material.WARPED_DOOR, StoneBrickDoor],
+  [Material.CRIMSON_DOOR, CobblestoneDoor],
+]);
+
+const DOOR_TYPES = new Map([
+  [Material.STONE_BRICKS, Material.WARPED_DOOR],
+  [Material.COBBLESTONE, Material.CRIMSON_DOOR],
+]);
+
+const doorCooldowns = new Set<Player>();
+
+// Hide the door
+registerEvent(PlayerInteractEvent, async (event) => {
+  if (event.action !== Action.LEFT_CLICK_BLOCK) return;
+  const block = event.clickedBlock;
+  if (!block) return;
+  if (doorCooldowns.has(event.player)) return;
+  const secretDoor = DOORS.get(block.type);
+  if (!secretDoor) return;
+  if (chanceOf(0.9)) return;
+
+  const door = block.blockData as Door;
+  let upper: Block;
+  let lower: Block;
+  if (door.half === Half.TOP) {
+    upper = block;
+    lower = block.getRelative(BlockFace.DOWN);
+  } else {
+    upper = block.getRelative(BlockFace.UP);
+    lower = block;
+  }
+
+  upper.setType(Material.AIR, false);
+  lower.setType(Material.AIR, false);
+  await wait(1, 'millis');
+  secretDoor.create(upper, {
+    hasLeftHinge: door.hinge === Hinge.LEFT,
+    facing: door.facing.toString(),
+  });
+  lower.type = upper.type;
+
+  playDoorCloseSound(block);
+  doorCooldowns.add(event.player);
+  await wait(0.5, 'seconds');
+  doorCooldowns.delete(event.player);
+});
+
+// Reveal the door
+const revealEvent = async (
+  event: PlayerInteractEvent,
+  data: { hasLeftHinge: boolean; facing: string },
+) => {
+  if (data.hasLeftHinge === undefined) return;
+  if (doorCooldowns.has(event.player)) return;
+  if (chanceOf(0.9)) return;
+  if (!event.clickedBlock) return;
+
+  const hinge = data.hasLeftHinge ? Hinge.LEFT : Hinge.RIGHT;
+  const block = event.clickedBlock;
+
+  // This deletes the customblock data
+  const blockBreakEvent = new BlockBreakEvent(block, event.player);
+  Bukkit.server.pluginManager.callEvent(blockBreakEvent);
```

Comment:
Our setBlock helper might be useful here. Apologies for missing documentation (aside of JSDoc).